File: TZCoordinateDemo/Objective-C Demo/ViewController.m
//
//  ViewController.m
//  TZCoordinateDemo
//
//  Created by Jason Wray on 3/27/17.
//  Copyright © 2019 Jason Wray. All rights reserved.
//

#import "ViewController.h"

//#import "NSTimeZone+Coordinate.h"
@import TimeZoneCoordinate;
@import MapKit;

@interface ViewController () <MKMapViewDelegate>

@property (weak, nonatomic) IBOutlet MKMapView *mapView;

@end

@implementation ViewController 

- (void)viewDidLoad {
    [super viewDidLoad];

    CLLocationCoordinate2D tzCoordinate = NSTimeZone.localTimeZone.coordinate;
    if (CLLocationCoordinate2DIsValid(tzCoordinate)) {
        [self.mapView setCenterCoordinate:tzCoordinate animated:YES];
    }

    NSLog(@"%@: %f, %f", NSTimeZone.localTimeZone, tzCoordinate.latitude, tzCoordinate.longitude);

    [self addPinsForAvailableTimeZones];

    if (@available(iOS 11.0, *)) {
        self.mapView.mapType = MKMapTypeMutedStandard;
    }
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
    MKPinAnnotationView *pin = (MKPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:@"pin"] ?: [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"pin"];
    pin.canShowCallout = YES;

    if ([annotation.title isEqualToString:NSTimeZone.localTimeZone.name]) {
        pin.pinTintColor = UIColor.systemGreenColor;
    }

    return pin;
}

- (void)addPinsForAvailableTimeZones {
    NSMutableArray *annotations = [NSMutableArray array];
    for (NSString *tzString in NSTimeZone.knownTimeZoneNames) {
        NSTimeZone *tz = [NSTimeZone timeZoneWithName:tzString data:nil];

        MKPointAnnotation *pin = [[MKPointAnnotation alloc] init];
        pin.coordinate = tz.coordinate;
        pin.title = tz.name;
        pin.subtitle = [NSString stringWithFormat:@"%.3f, %.3f", tz.coordinate.latitude, tz.coordinate.longitude];

        [annotations addObject:pin];
    }

    [self.mapView addAnnotations:annotations];
}

@end

File: TZCoordinateDemo/Timezone Coordinate Demo/ViewController.swift
//
//  ViewController.swift
//  Timezone Coordinate Demo
//
//  Created by Jason Wray on 8/25/19.
//  Copyright © 2019 Jason Wray. All rights reserved.
//

import TimeZoneCoordinate
import MapKit

class ViewController: UIViewController {
    @IBOutlet var mapView: MKMapView!

    override func viewDidLoad() {
        super.viewDidLoad()

        if let coordinate = TimeZone.current.coordinate {
            mapView.centerCoordinate = coordinate
            print("\(TimeZone.current): \(NSString(format: "%.3f, %.3f", coordinate.latitude, coordinate.longitude))")
        }

        addPinsForAvailableTimeZones()

        if #available(iOS 11.0, *) {
            mapView.mapType = .mutedStandard
        }
    }

    func addPinsForAvailableTimeZones() {
        var annotations: [MKPointAnnotation] = []
        for timeZoneIdentifier in TimeZone.knownTimeZoneIdentifiers {
            guard let timezone = TimeZone(identifier: timeZoneIdentifier) else { return }

            if let coordinate = timezone.coordinate {
                let annotation = MKPointAnnotation()
                annotation.coordinate = coordinate
                annotation.title = timezone.identifier
                annotation.subtitle = NSString(format: "%.3f, %.3f", coordinate.latitude, coordinate.longitude) as String

                annotations.append(annotation)
            }
        }

        mapView.addAnnotations(annotations)
    }
}

extension ViewController: MKMapViewDelegate {
    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        var pin: MKPinAnnotationView

        if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: "pin") as? MKPinAnnotationView {
            pin = annotationView
        } else {
            pin = MKPinAnnotationView(annotation: annotation, reuseIdentifier: "pin")
        }

        pin.canShowCallout = true

        if annotation.title == TimeZone.current.identifier {
            pin.pinTintColor = .systemGreen
        }

        return pin;
    }
}
